packages/babel-core/src/config/option-manager.js

```diff
@@ -12,11 +14,11 @@ import clone from "lodash/clone";
 import { loadPlugin, loadPreset, loadParser, loadGenerator } from "./loading/files";
 
 type MergeOptions = {
-  type: "arguments"|"options"|"preset",
-  options?: Object,
+  +type: "arguments"|"options"|"preset",
```

Comment:
This is property covariance. In this case it essentially allows two types that are mostly-compatible to play nicely together. Take this case:

```
type ConfigItem = {
  type: "options"|"arguments",
};

type MergeOptions = {
  type: "arguments"|"options"|"preset",
};

var item: ConfigItem = {
  type: "options",
};

var merge: MergeOptions = item;
```

This will fail, because `merge.type = "preset";` would cause `item` to no longer match its type properly, since `ConfigItem` objects cant be have type `preset`.

By adding the `+`, you are basically saying "don't let me reassign this", so

```
var item: ConfigItem = {
  type: "options",
};

var merge: MergeOptions = item;
```

will work, but it will throw an error if you try to reassign `merge.type`.